### PR TITLE
DO NOT MERGE - NEWRELIC-5296: reduce cognitive complexity of lib/shim/message-shim.js

### LINKED ISSUE #981: Research replacing recordConsume instrumentation wrapping with record
We found a bug in `recordConsume` #977 and we made a quick fix.  This ticket will serve as a research/prototype to investigate whether or not we can just use `this.record` within `recordConsume` instead of `this.wrap`.  It looks like for this method we made up a `messageHandler` spec key which executes after a callback finishes or promise resolves and we copy parameters to the segment.  I'm wondering if we can just use `spec.after` to achieve this?  But also since this API is exposed would it be a breaking change?

### PR DESCRIPTION
**Pull request by jordigh**
## Links

* NEWRELIC-5296

## Details

Commit messages:

*NEWRELIC-5296: (recordConsume) reduce its cognitive complexity

Also closes #981, since it removes the biggest use of `shim.wrap` with
`this.record`. The resulting function is much simpler!

*NEWRELIC-5296: (message-shim) break up recordSubscribedConsume into functions

This one is kind of complicated and it's even more complicated to
simplify it. The only way I could see to make it better was to break
it up into smaller functions of acceptable complexity.

*NEWRELIC-5296: (message-shim) remove eslint cognitive complexity exception

File now lints properly without it.

### CODE REVIEW

File: lib/shim/message-shim.js [hunk 1]
```diff
@@ -589,43 +516,13 @@ function recordSubscribedConsume(nodule, properties, spec) {
     promise: false
   })
 
-  // Make sure our spec has what we need.
```
Comment (jordigh) [inline lib/shim/message-shim.js:592]: Maybe I should bring these back. I got rid of them because I figured it was better to just error out completely if you gave this function the wrong arguments than to silently proceed and write something to debug output.

It's just a complete error of custom instrumentation to supply the wrong arguments here, so it's not exactly like getting rid of these two checks breaks user code.

File: lib/shim/message-shim.js [hunk 2]
```diff
@@ -356,108 +354,37 @@ function recordConsume(nodule, properties, spec) {
...
-                return cb.apply(this, arguments)
-              }
-            }
-          })
-        }
+  return this.record(nodule, properties, function consumeRecorder(shim, fn, fnName, args) {
+    if (shim.isFunction(spec)) {
+      spec = spec.apply(this, arguments)
+    } else {
+      spec = { ...DEFAULT_SPEC, ...spec }
```
Comment (bizob2828) [inline lib/shim/message-shim.js:364]: what if you just change `shim.setDefaults` to do this?  that way it'll encapsulated in API

File: lib/shim/message-shim.js [hunk 3]
```diff
@@ -356,108 +354,37 @@ function recordConsume(nodule, properties, spec) {
...
+          shim.copySegmentParameters(segment, msgProps.parameters)
         }
+        return val
       }
+    }
 
-      return ret
+    const name = _nameMessageSegment(shim, spec, shim._metrics.CONSUME)
+    return {
+      name: name,
```
Comment (bizob2828) [inline lib/shim/message-shim.js:385]: nit: you don't need to define name as both key and value

File: lib/shim/message-shim.js [hunk 4]
```diff
@@ -356,108 +354,37 @@ function recordConsume(nodule, properties, spec) {
...
-                if (getParams && msgProps && msgProps.parameters) {
-                  shim.copySegmentParameters(segment, msgProps.parameters)
-                }
-
-                return cb.apply(this, arguments)
-              }
-            }
-          })
-        }
+  return this.record(nodule, properties, function consumeRecorder(shim, fn, fnName, args) {
```
Comment (bizob2828) [inline lib/shim/message-shim.js:360]: the interesting net of this is it will capture CLM attrs now for these functions, not sure if it's relevant but was just tracing this through

File: lib/shim/message-shim.js [hunk 5]
```diff
@@ -356,108 +354,37 @@ function recordConsume(nodule, properties, spec) {
...
-              shim.copySegmentParameters(segment, msgProps.parameters)
-            }
-            return res
-          })
+    }
+    const messageHandler = spec.messageHandler
+    const getParams = shim.agent.config.message_tracer.segment_parameters.enabled
+
+    if (messageHandler) {
+      spec.after = function afterPromise({ shim, val, segment }) {
```
Comment (bizob2828) [inline lib/shim/message-shim.js:374]: this really cuts down on so much code, nice work 👏🏻

File: lib/shim/message-shim.js [hunk 6]
```diff
@@ -649,132 +546,176 @@ function recordSubscribedConsume(nodule, properties, spec) {
...
-      msgDescDefaults.queue = queue
+      spec.queue = queue
     }
 
-    return function wrapConsumer(shim, consumer, cName) {
-      if (!shim.isFunction(consumer)) {
-        return consumer
-      }
+    if (consumerIdx !== null) {
+      args[consumerIdx] = shim.wrap(args[consumerIdx], function wrapConsumer(shim, consumer, name) {
```
Comment (bizob2828) [inline lib/shim/message-shim.js:589]: interesting that you were able to cut out a lot of code by checking things above and then just returning this simple function

File: lib/shim/message-shim.js [hunk 7]
```diff
@@ -356,108 +354,37 @@ function recordConsume(nodule, properties, spec) {
...
-        if (resHandler) {
-          ret = ret.then(function interceptValue(res) {
-            const msgProps = resHandler.call(this, shim, fn, fnName, res)
-            if (getParams && msgProps && msgProps.parameters) {
-              shim.copySegmentParameters(segment, msgProps.parameters)
-            }
-            return res
-          })
+    }
+    const messageHandler = spec.messageHandler
```
Comment (bizob2828) [inline lib/shim/message-shim.js:370]: this is so weird, I don't see any of our instrumentation using this: sns, sqs, nor amqp.  We have tests for it though and technically this is a pubilc API but wow

File: lib/shim/message-shim.js (hunk 5 above)
Comment (bizob2828) [reply lib/shim/message-shim.js:374]: this is a breaking change right? we've changed the signature to be object based.  I know we updated our code to now parse the object but customers may be unhappy.  I will take with semver:major and maybe we can release a semver major soon with this, my refactor of registering multiple hooks and #1520

**Review by bizob2828 — COMMENTED** [review]
Overall great work making sense of this and reducing a ton of code.  I added `semver:major`. I think this is a breaking change because of the after signature change.  Now that we have 3 semver major PRs opened maybe we release that in a few weeks?

File: lib/shim/message-shim.js (hunk 2 above)
Comment (jordigh) [reply lib/shim/message-shim.js:364]: I kind of deliberately didn't use `shim.setDefaults`, because I thought it was rather pointless function. It predates the spread operator. It does make a slight difference because it starts from `Object.create(null)` though. Do you really think it's worth to go back to it?

File: lib/shim/message-shim.js (hunk 4 above)
Comment (jordigh) [reply lib/shim/message-shim.js:360]: Oh, yes, maybe I should add some tests for CLM now that it goes through a common codepath?

File: lib/shim/message-shim.js (hunk 5 above)
Comment (jordigh) [reply lib/shim/message-shim.js:374]: So, we do not publicly document what should go into`spec.after`, other than saying that it's a callback function. I guess this is a breaking change but it's such an obscure part of our API that I wonder if anyone even uses it.

However, if we already have another reason to introduce a breaking change, this is a good place to do it.

File: lib/shim/message-shim.js (hunk 2 above)
Comment (bizob2828) [reply lib/shim/message-shim.js:364]: I'm suggesting you change shim.setDefaults to spread to encapsulate this in a function and not just one of spreading in some places and not in others

Comment (jordigh) [reply lib/shim/message-shim.js:364]: Okay, will do.

File: lib/shim/message-shim.js (hunk 5 above)
Comment (bizob2828) [reply lib/shim/message-shim.js:374]: shim API is a grey area because we expose it. Even if we don't document it, it's publicly available. We should just ship this with our next semver major so hold off on merging. We discussed in standup today that we will be making a semver major release shortly(maybe this week)

File: lib/shim/message-shim.js (hunk 7 above)
Comment (jordigh) [reply lib/shim/message-shim.js:370]: I can't find anyone on Github using it either, but maybe we have some secret fans out there.

Comment (bizob2828) [reply lib/shim/message-shim.js:370]: If we're making this a semver major should we just remove?

Comment (jordigh) [reply lib/shim/message-shim.js:370]: I don't see a reason to break more than we have to. It's kind of convenient to to keep it around. Would be more work to remove it: would have to update our docs and examples.

Comment (jordigh) [reply lib/shim/message-shim.js:370]: Wait, we do use it ourselves: we use it for amqplib.

File: lib/shim/message-shim.js (hunk 6 above)
Comment (jordigh) [reply lib/shim/message-shim.js:622]: I accidentally cut too much 'cause I got confused by the callback of the callback. I just put a bit back in.

File: lib/shim/message-shim.js [hunk 8]
```diff
@@ -361,34 +361,67 @@ function recordConsume(nodule, properties, spec) {
...
+ *  The user-supplied specification for this recording method's interface.
+ */
+function _wrapCallBackWithMessageHandler(shim, messageHandler, args, spec) {
+  const getParams = shim.agent.config.message_tracer.segment_parameters.enabled
+  let cbArgs
+  let origCb
+  let origCbName
+  const cbIdx = shim.normalizeIndex(args.length, spec.callback)
+  if (cbIdx !== null) {
+    // wrap the callback to save the args
```
Comment (bizob2828) [inline lib/shim/message-shim.js:405]: this is hard to follow but comparing from before, it was doing shim.bindCallbackSegment. do we no longer need this because it's using shim.record above?  Also the wrapped callback is confusing to me, can you explain why it's only storing the original args as a reference then calling the original callback?  Also, not understanding the defineProperty of name to keep function the same

Comment (jordigh) [reply lib/shim/message-shim.js:405]: Okay, yeah, so maybe this whole thing didn't actually reduce cognitive complexity. Here's the situation:

We have to wrap a function (let's call it `getMessage`, like in the `newrelic-examples` case) that calls a callback (let's call it `callback`). The old method was doing this by calling `shim.wrap` inside a `shim.wrap`. That wraps both `getMessage` and `callback`. This works but it ends up duplicating a lot of the functionality of `spec.record` and is complicated to follow.

But new method now calls `shim.record` and uses `spec.after` to call `getMessage`'s callback. This almost works. The callback gets called by `spec.after` at the right time. The only problem is that `spec.after`, like its name implies, is called *after* `callback`. So first `callback` is called and then `spec.after` is called which in this case does stuff like putting the parameters into the segment. Defining the appropriate segments and their attributes for both `getMessage` and `callback` is now all handled by the `shim.record` function.

The problem is that because `spec.after` is called after, we don't have `callback`'s arguments. Those were invoked when `getMessage` called `callback`.

So I patched this by doing a minimal, ad-hoc wrapping of `callback` that all it does is grab its arguments, and not all of `shim.wrap`'s other things. Because I don't want to change the way segments are named (that would break a test in `test/versioned/amqplib/callback.tap.js`), I also make sure the wrapped callback has the same function name as `callback`.

What do you think? Maybe the whole thing is now just simply complicated in a different way and this was no improvement at all.

Comment (bizob2828) [reply lib/shim/message-shim.js:405]: this is not working as you expect.  the wrapped callback gets called after the .after function below.  this is because of ordering in event loop.  I also don't think this is the right way to approach this. I'm still digging into the _right_ way but it's getting very confusing 😄 .  Honestly at this point I'm wondering if we just keep this open and go forward with the major release.

**Comment by bizob2828** [comment]
I think we should close this and reopen with just reducing cognitive complexity, not refactoring to use shim.record.  This change is too risky and I'm struggling to map old functionality with new

**Comment by bizob2828** [comment]
I'm going to close this. we can revisit at a later date
